sn_client/src/files/upload.rs

```diff
@@ -186,210 +218,170 @@ impl FilesUpload {
+                    );
+                    if cost.2.cost != NanoTokens::zero() {
+                        pending_to_pay.push((chunk_info, cost));
+                    } else if for_failed_chunks {
                         self.send_event(FileUploadEvent::Uploaded(ChunkAddress::new(
                             chunk_info.name,
-                        )))
-                        .await?;
+                        )));
+                        uploaded += 1;
+                    } else {
+                        self.send_event(FileUploadEvent::AlreadyExistsInNetwork(
+                            ChunkAddress::new(chunk_info.name),
+                        ));
+                        skipped += 1;
                     }
-                    (chunk_info, Err(err)) => {
-                        warn!("Failed to upload a chunk: {err}");
-                        self.send_event(FileUploadEvent::FailedToUpload(ChunkAddress::new(
-                            chunk_info.name,
-                        )))
-                        .await?;
-                        // store the failed chunk to be retried later
-                        self.failed_chunks.insert(chunk_info);
+                }
+                TaskResult::MakePaymentsOK((
+                    (storage_cost, royalty_fees, new_balance),
+                    reply_list,
+                )) => {
+                    trace!("Paid {} chunks, with {storage_cost:?} store_cost and {royalty_fees:?} royalty_fees, and new_balance is {new_balance:?}",
+                        reply_list.len());
+                    sequential_payment_fails = 0;
+                    for (chunk_info, _) in reply_list.iter() {
+                        let _ = on_going_pay_for_chunk.remove(&chunk_info.name);
+                    }
+                    pending_to_upload.extend(reply_list);
+
+                    self.upload_storage_cost = self
+                        .upload_storage_cost
+                        .checked_add(storage_cost)
+                        .ok_or(ClientError::TotalPriceTooHigh)?;
+                    self.upload_royalty_fees = self
+                        .upload_royalty_fees
+                        .checked_add(royalty_fees)
+                        .ok_or(ClientError::TotalPriceTooHigh)?;
+                    self.upload_final_balance = new_balance;
+                    self.send_event(FileUploadEvent::PayedForChunks {
+                        storage_cost,
+                        royalty_fees,
+                        new_balance,
+                    });
+                }
+                TaskResult::UploadChunksOK(xor_name) => {
+                    trace!("Upload task uploaded chunk {xor_name:?}");
+                    let _ = on_going_uploadings.remove(&xor_name);
+                    uploaded += 1;
+                    self.send_event(FileUploadEvent::Uploaded(ChunkAddress::new(xor_name)));
+                }
+                TaskResult::ErrorEntries(error_list) => {
```

Comment:
Can we add a comment here stating that, it is mandatory to throw an error inside here?
Since we don't have separate error variant for each task (say upload_chunks), we might not be clearing up `on_going_uploadings` if we failed to upload a chunk.
And the loop would not progress as `on_going_uploadings.len() < batch_size` would never happen. This will block the other tasks too.

Reply:
Throwing an error breaks the loop.